File: src/main/java/org/trp/shincolle/network/C2SFormationActionPayload.java
package org.trp.shincolle.network;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import org.trp.shincolle.Shincolle;

import java.util.Optional;
import java.util.UUID;

public record C2SFormationActionPayload(int action, int param1, int param2, String paramString, Optional<UUID> paramUUID) implements CustomPacketPayload {
    public static final Type<C2SFormationActionPayload> TYPE = new Type<>(ResourceLocation.fromNamespaceAndPath(Shincolle.MODID, "c2s_formation_action"));

    public static final StreamCodec<FriendlyByteBuf, C2SFormationActionPayload> STREAM_CODEC = StreamCodec.composite(
            ByteBufCodecs.VAR_INT, C2SFormationActionPayload::action,
            ByteBufCodecs.VAR_INT, C2SFormationActionPayload::param1,
            ByteBufCodecs.VAR_INT, C2SFormationActionPayload::param2,
            ByteBufCodecs.STRING_UTF8, C2SFormationActionPayload::paramString,
            ByteBufCodecs.optional(UUIDUtil.STREAM_CODEC), C2SFormationActionPayload::paramUUID,
            C2SFormationActionPayload::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    private static class UUIDUtil {
        public static final StreamCodec<FriendlyByteBuf, UUID> STREAM_CODEC = new StreamCodec<>() {
            @Override
            public UUID decode(FriendlyByteBuf buffer) {
                return buffer.readUUID();
            }

            @Override
            public void encode(FriendlyByteBuf buffer, UUID value) {
                buffer.writeUUID(value);
            }
        };
    }
}

File: src/main/java/org/trp/shincolle/network/C2SFormationActionPayload.kt
package org.trp.shincolle.network

import com.mojang.datafixers.util.Function5
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation
import org.trp.shincolle.Shincolle
import java.util.*

@JvmRecord
data class C2SFormationActionPayload(
    val action: Int,
    val param1: Int,
    val param2: Int,
    val paramString: String?,
    val paramUUID: Optional<UUID?>
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload?> {
        return TYPE
    }

    private object UUIDUtil {
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf, UUID> = object : StreamCodec<FriendlyByteBuf, UUID> {
            override fun decode(buffer: FriendlyByteBuf): UUID {
                return buffer.readUUID()
            }

            override fun encode(buffer: FriendlyByteBuf, value: UUID) {
                buffer.writeUUID(value)
            }
        }
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<C2SFormationActionPayload?> =
            CustomPacketPayload.Type<C2SFormationActionPayload?>(
                ResourceLocation.fromNamespaceAndPath(
                    Shincolle.MODID,
                    "c2s_formation_action"
                )
            )

        val STREAM_CODEC: StreamCodec<FriendlyByteBuf?, C2SFormationActionPayload?> =
            StreamCodec.composite<FriendlyByteBuf?, C2SFormationActionPayload?, Int?, Int?, Int?, String?, Optional<UUID?>?>(
                ByteBufCodecs.VAR_INT,
                C2SFormationActionPayload::action,
                ByteBufCodecs.VAR_INT,
                C2SFormationActionPayload::param1,
                ByteBufCodecs.VAR_INT,
                C2SFormationActionPayload::param2,
                ByteBufCodecs.STRING_UTF8,
                C2SFormationActionPayload::paramString,
                ByteBufCodecs.optional<FriendlyByteBuf, UUID>(UUIDUtil.STREAM_CODEC),
                C2SFormationActionPayload::paramUUID,
                Function5 { action: Int?, param1: Int?, param2: Int?, paramString: String?, paramUUID: Optional<UUID?> ->
                    C2SFormationActionPayload(
                        action!!,
                        param1!!,
                        param2!!,
                        paramString,
                        paramUUID
                    )
                }
            )
    }
}
